include/nbl/builtin/hlsl/subgroup/basic_portability.hlsl

```diff
@@ -0,0 +1,127 @@
+	
+	uint ElectedSubgroupInvocationID() {
+		return BroadcastFirst<uint>(InvocationID());
+	}
+	
+	uint ElectedLocalInvocationID() {
+		return BroadcastFirst<uint>(gl_LocalInvocationIndex);
+	}
+
+	template<typename T>
+	T Broadcast(T value, uint invocationId) {
+		return WaveReadLaneAt(value, invocationId);
+	}
+
+	// REVIEW: Masks don't seem to be available in Wave Ops
+	uint64_t EqMask() { // return type for these must be 64-bit
+		// REVIEW: I think the spec says that the subgroup size can potentially go up to 128 in which case uint64 won't work properly...
+		//  Should we somehow block shaders from running if subgroup size > 64?
+		//  Use uint64_t2 ?
+		return 0x1u << InvocationID();
+	}
+	
+	uint64_t GeMask() {
+		return 0xffffffffu << InvocationID();
+	}
+
+	uint64_t GtMask() {
+		return GeMask()<<1;
+	}
+
+	uint64_t LeMask() {
+		return ~GtMask();
+	}
+
+	uint64_t LtMask() {
+		return ~GeMask();
+	}
+
+	// REVIEW: Should we define the Small/LargeSubgroupXMask and Small/LargeHalfSubgroupSizes?
+
+	// WAVE BARRIERS
+
+	// REVIEW: Review everything related to subgroup barriers and SPIR-V injection
+	// REVIEW: Should we check #ifdef NBL_GL_KHR_shader_subgroup_XYZ before applying spirv_subgroupBarriers? Maybe add GroupBarriers if not defined?
+
+	[[vk::ext_instruction(/* OpControlBarrier */ 224)]] // https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpControlBarrier
+	void spirv_subgroupBarrier(uint executionScope, uint memoryScope, uint memorySemantics);
+
+	// REVIEW: Should we name the Barriers with the Subgroup prefix just to make it clearer when calling?
+	// REVIEW: Proper Memory Semantics!! Link here: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Memory_Semantics_-id-
+	// REVIEW: Need advice on memory semantics. Would think SubgroupMemory(0x80) | AcquireRelease(0x8) is the correct bitmask but SubgroupMemory doesn't seem to be supported as Vulkan storage class
+	
+	void Barrier() {
+		// https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_scope_id
+		// Subgroup scope is number 3, both for execution and memory
+
+		// https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id
+		// By providing memory semantics None we do both control and memory barrier as is done in GLSL
+		
+		// REVIEW: After testing it seems that this doesn't work at all even if we provide Subgroup execution scope and it 
+		// compiles as `OpControlBarrier %uint_3 %uint_3 %uint_264` which is supposed to be correct. What seems to actually 
+		// happen is that this executes as a workgroup scoped barrier which can break uniform barrier calls and produce UB.
+		// Since subgroups execute in lockstep we could maybe just avoid calling this?
+		//spirv_subgroupBarrier(3, 3, 0x8 | 0x80);
+	}
+
+	[[vk::ext_instruction(/* OpMemoryBarrier */ 225)]] // https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpControlBarrier
+	void spirv_subgroupMemoryBarrierShared(uint memoryScope, uint memorySemantics);
```

Comment:
leave some docs around
```
0 = cross device
1 = device
2 = workgroup
3 = subgroup
4 = invocation
5 = queue family
6 = shadercall
```

Reply:
btw do you mind making enums or constants for these?

Reply:
don't call it `spirv_subgroupMemoryBarrierShared`, call it `spirv_memoryBarrier` because all 8 barrier types (`{subgroup,global} x {global,shared,buffer,image}`) get implemented in terms of this opcode

Reply:
enums/docs/defines still useful for functions in  new `/hlsl/spirv/...` headers

Reply:
including the SPIRV-Headers in the HLSL works!

https://godbolt.org/z/G45ErGx1h

also slapping the enums into the `[]` attributes to be used as integral constants also works!

@AnastaZIuk will copy over the SPIR-H header with CMake to `nbl/builtin/hlsl/spirv/`